csrc/multidevice/utils.cpp

```diff
@@ -151,6 +217,116 @@ void insertReshardings(Fusion* fusion) {
   }
 }
 
+void insertShardedAxisReordering(Fusion* fusion) {
+  auto exprs = fusion->exprs();
+  std::vector<Expr*> reshard_exprs;
+  for (auto expr : exprs) {
+    if (isResharding(expr)) {
```

Comment:
could we add another check at this point to verify whether the expr needs axis reordering?
I understand that this check is somwhow done later by calling `isOutermostAllocatedId`, but it would be nice to have one method returning a bool wether an I/O sharded axis is non-outermost.

This way, we could add an assert in `isLowerableToCommunication` to make sure at the symbolic level that there is no communications on non-outermost domain.